Refactor and add JSDoc support

dcdunkan opened the pull request:
Re-wrote the whole thing once again. Fixed few mistakes, and made new mistakes.

- Pretty issue reports and neat summary.
- Ability to check for broken links in a JS/TS module.
- Groupable links (plugin?). Some types of links needs to be taken care of in a special way.
- CLIs have been improved and split into a different file.
- WAY better redirection validator.

Closes #1

dcdunkan committed:
refactor for TSDocs support

dcdunkan committed:
chore: cleanup

dcdunkan committed:
fmt

dcdunkan committed:
refactor: group links

dcdunkan committed:
oops

dcdunkan committed:
feat: check TSDoc

dcdunkan committed:
feat: pretty issues

dcdunkan committed:
support for non-tty env

dcdunkan committed:
fix: pretty print issues

dcdunkan committed:
refactor: re-arrange stuff

dcdunkan committed:
chore: bump deps

dcdunkan committed:
chore: bump deps

dcdunkan committed:
ref: separate cli parts

dcdunkan committed:
random stuff

dcdunkan committed:
random stuff again

dcdunkan committed:
finalize?

dcdunkan committed:
a lot of stuffgt

* grouping for renderable github files such as directory readmes, etc.
* more arguments (and defaults) to website CLI.
* fixed redirection check for transformed links.
* fix references here and there.
* updated readme
* better errors
* refactored and improved minor stuff

dcdunkan committed:
Fix and reorder Deno permissions

dcdunkan committed:
bump deps

dcdunkan committed:
include deps/ in deno check

dcdunkan committed:
Sync with changes from #7: Consecutive forward slashes

dcdunkan committed:
fix return

dcdunkan committed:
fix merge conflicts

dcdunkan committed:
no decoding

dcdunkan committed:
fix all merge conflicts

dcdunkan committed:
remove 'reading' logging for directories

dcdunkan committed:
mention about github api token

dcdunkan committed:
improve the paragraph

dcdunkan committed:
unquote

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
-A tool for going through
-[grammY documentation source](https://github.com/grammyjs/website) and reporting
-broken links.
+A tool for going through grammY documentation source and other repositories' JSDocs, and reporting broken links and missing
```
```suggestion
A tool for going through grammY documentation source and other repositories’ JSDocs, and reporting broken links and missing
```

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
-To use the tool locally, clone the
-[grammY website repository](https://github.com/grammyjs/website) and go to the
-`site/docs/` directory and run the following:
+To check for broken links in [grammY's documentation source](https://github.com/grammyjs/website), clone the website repository
```
```suggestion
To check for broken links in the [grammY documentation](https://github.com/grammyjs/website), clone the website repository
```

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
+
+Supported arguments:
+
+- `--clean-url`: Defaults to `false`. Vitepress's clean URL config option makes sure that the references to other local files
```
```suggestion
- `--clean-url`: Defaults to `false`. VitePress’ clean URL config option makes sure that the references to other local files
```

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
+Supported arguments:
+
+- `--clean-url`: Defaults to `false`. Vitepress's clean URL config option makes sure that the references to other local files
+  doesn't end with an extension. Specify this option if the configuration used in the website repository has this option enabled.
```
```suggestion
  doesn’t end with an extension. Specify this option if the configuration used in the website repository has this option enabled.
```

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
+[supported markup](https://github.com/github/markup/blob/master/README.md#markups) file, because there is an anchor. When normally
+fetched, their responses does not contain the actual content of that file, because the content is lazily loaded. So for checking
+if the anchor is present, we fetch the content using the GitHub Content REST API. And that is why we need a GitHub access token.
+If you only have a few such links, then you most likely don't need the token.
```
```suggestion
If you only have a few such links, then you most likely don’t need the token.
```

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
+
+When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here's why:
+There might be links like https://github.com/OWNER/REPO/tree/dir#anchor or https://github.com/OWNER/REPO/blob/file.rst#anchor in
+the documentation. These types of links are assumed to be pointing to a GitHub repository's directory README or a
```
```suggestion
the documentation. These types of links are assumed to be pointing to a GitHub repository’s directory README or a
```

rojvv commented on README.md:
```diff
@@ -1,16 +1,53 @@
+**TLDR;** Set the environment variable `GITHUB_TOKEN` if documents contains a lot of links to GitHub markup (README, for example)
+files. A fine-grained token with the default set of permissions will do.
+
+When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here's why:
```
```suggestion
When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here’s why:
```

dcdunkan committed:
thanks @roj1512

Co-authored-by: Roj <rojvv@icloud.com>

quadratz commented:
Please correct me if I'm mistaken, but if the link checker requires a GitHub token, it is designed to be used specifically with GitHub Action, correct? Unless we are comfortable hard coding the token in our code, users will need to utilize their own GitHub token and run it on their individual PCs.

```console
➜  deno run -r --allow-env --allow-net --allow-read website_cli.ts ./                                          

┃ Gentle reminder: It is recommended to set the GITHUB_TOKEN environment variable
┃ if there are GitHub repository file links in the documents.

reading README.md
It's a renderable Github file with an anchor. Resolving using Github API...
error: Uncaught Error: Unauthorized
  if (!response.ok) throw new Error(response.statusText);
                          ^
```

dcdunkan commented:
Thanks for catching that. That was a mistake, that part is wrong. Will fix that. I realized that I actually never tested it without a token after I completed it.

Link checker doesn't "require" the token unless there are a lot of such links. Unauthorized requests have 60 req/h limit. And no, it is not designed specifically to use with GitHub Actions. But for a repository like our website's, it is reasonable to assume that we are gonna need a token, because we can't simply predict the number of such links (even tho we currently only have 1 link, and only two request are made, one for branches and the other for the rendered file content).

Here it was supposed to be leave the reminder and fetch it without the token. But I got lost.

dcdunkan committed:
Fix unauthorized Github requests

dcdunkan commented:
Can you try it now? (without a token)

dcdunkan committed:
Fix issue titles and issue summary templates

dcdunkan commented:
> ```
> ➜  deno run -r --allow-env --allow-net --allow-read website_cli.ts ./
> ```

You're gonna need `--clean-url` :D

dcdunkan committed:
fmt

dcdunkan committed:
Rename ts_doc to tsdoc

dcdunkan committed:
check file type

dcdunkan committed:
Fix README

dcdunkan committed:
l

quadratz commented:
Yup. It's working now. Let me do some testing

Commit:
Not real URL

Commit:
Merge branch 'refactor' of github.com:grammyjs/link-checker into refactor

Commit:
wip: add some test files

quadratz reviewed (commented):
There should be at least 11 errors, not 5.

```console
➜  deno run -r --allow-env --allow-net --allow-read website_cli.ts --clean-url test/clean-url/

┃ Gentle reminder: It is recommended to set the GITHUB_TOKEN environment variable
┃ if there are GitHub repository file links in the documents.

reading test\clean-url\bar.md
reading test\clean-url\foo.md
reading test\clean-url\id\bar.md
reading test\clean-url\id\foo.md
reading test\clean-url\id\guide\bar.md
reading test\clean-url\id\guide\foo.md
reading test\clean-url\id\guide\README.md
reading test\clean-url\id\README.md
reading test\clean-url\invalid\bar.md
reading test\clean-url\invalid\es\foo.md
reading test\clean-url\invalid\es\guide\foo.md
reading test\clean-url\invalid\foo.md
reading test\clean-url\README.md
Found 5 issues in the documentation:

┌────────────────────────┬───┐
│          Missing files │ 5 │
├────────────────────────┼───┤
│                  Total │ 5 │
└────────────────────────┴───┘

test\clean-url\invalid\bar.md (4)
— The linked file ../id/bazzz.md does not exist.
— The linked file ./lost-file.md does not exist.
— The linked file ./es/lost-file.md does not exist.
— The linked file ./es/guide/lost-file.md does not exist.

test\clean-url\invalid\foo.md (1)
— The linked file ./id/bar.md does not exist.
```

Commit:
Catch multiple slashes in path

dcdunkan commented:
> There should be at least 11 errors, not 5.

It'd be easier to fix if could you post all the expected ones?

Commit:
Add more test

Commit:
Fix "./#my-anchor"

Commit:
False negative

quadratz commented:
@dcdunkan Just one more left. This url https://deno.land/x/grammy@v0.00.0 should return 404

quadratz commented:
`200 OK`? Really?! 🤯 

![image](https://github.com/grammyjs/link-checker/assets/74030149/5264a714-0af6-4f74-b002-1c577403bf8f)

![image](https://github.com/grammyjs/link-checker/assets/74030149/0143bb94-52a3-4a0b-a137-8e7726ff4dff)

dcdunkan committed:
should be fixed at the right place

dcdunkan commented:
Yeah, I also checked. Consider opening an issue at denoland/website_feedback. This shouldn't be adjusted at our side.

dcdunkan committed:
dont export things we dont need

quadratz reviewed (approved):
LGTM. The link-checker passed all of my tests, and I also tested it on the grammy/website repository. Thank you for your hard work.

quadratz commented:
Should we remove the test folder, @dcdunkan?

dcdunkan commented:
> LGTM. The link-checker passed all of my tests, and I also tested it on the grammy/website repository. Thank you for your hard work.

Awesome, thank you so much for testing and catching my mistakes.

dcdunkan commented:
> Should we remove the test folder, @dcdunkan?

Yeah. Writing tests for this was in my head when I started, then I got lazy. But now that you have already did this, I'd like to write some actual tests. I'll take care of that now. Might need another review for that too :)

quadratz commented:
Aye aye capt!

Commit:
Remove test files

Commit:
Fix redirect

"The link https://deno.com/manual/runtime/workers was redirected to https://deno.land/manual@v1.35.1/runtime/workers."

dcdunkan commented:
@CikiMomogi Please revert 35887f1. That's not a valid redirection.

They have plans about merging deno.com and deno.land. They first redirected certain links from land to com and then they didn't change anything for some time, now this too, com to land. Since land's source code is archived now and com has always been closed source, any changes could take place, that we don't know anything about. @KnorpelSenf and I talked about this and have decided to do nothing about them until they settle for enough time. That's why this is a invalid redirection.

When I fixed the links in website, there was this redirection:
https://deno.land to https://deno.com/runtime. And for a while they stayed like that, and they changed it back, no redirections. Then again, they now redirects deno.com/runtime to deno.land lol. So, skip it.

Reference about their merging plans: https://github.com/denoland/website_feedback/issues/34

Commit:
Revert "Fix redirect"

This reverts commit 35887f1ae0fdcca091afe9943da7e65fe4369b2d.

quadratz commented:
I suggest keeping it to deno.land for the time being since we're not entirely sure when the link will be fixed.

dcdunkan commented:
I don't agree to that, sorry. If some other intended deno.com link does a valid redirection to deno.com itself, then that would be an invalid redirection since we're expecting deno.land. So, I'd rather prefer fixing these issues in the website repository; or as discussed, wait a while and then do something about it.

quadratz commented:
@dcdunkan, is this PQ good to merge? I'm pretty sure it'll fix the anchor-check error in the current link-checker.

KnorpelSenf commented:
I think the whole deno.com/deno.land thing has settled now, too, so we can finish this.

dcdunkan committed:
remove unwanted redrirection validations

dcdunkan committed:
chore: update deps

dcdunkan committed:
fix: google documentation anchors detecting

dcdunkan committed:
chore: random final touches

dcdunkan commented:
@quadratz @KnorpelSenf Sorry about the late response. Nothing important changed in the last 4 commits. So, I'll just merge it now. I hope @quadratz can send the update PQ in /website? Thanks for reviewing and catching my mistakes!

(I was lazy again, so I didn't add any tests)
